src/window.cpp

```diff
@@ -2901,19 +2897,28 @@ static void MouseLoop(MouseClick click, int mousewheel)
 			case MC_LEFT:
 			case MC_DOUBLE_LEFT:
 				DispatchLeftClickEvent(w, x - w->left, y - w->top, click == MC_DOUBLE_LEFT ? 2 : 1);
-				break;
+				return;
 
 			default:
 				if (!scrollwheel_scrolling || w == NULL || w->window_class != WC_SMALLMAP) break;
 				/* We try to use the scrollwheel to scroll since we didn't touch any of the buttons.
 				 * Simulate a right button click so we can get started. */
 				FALLTHROUGH;
 
-			case MC_RIGHT: DispatchRightClickEvent(w, x - w->left, y - w->top); break;
+			case MC_RIGHT:
+				DispatchRightClickEvent(w, x - w->left, y - w->top);
+				return;
 
-			case MC_HOVER: DispatchHoverEvent(w, x - w->left, y - w->top); break;
+			case MC_HOVER:
+				DispatchHoverEvent(w, x - w->left, y - w->top);
+				break;
 		}
 	}
+
```

Comment:
It feels a bit weird that it depends if it was a `return` or `break` in the case above, especially as that is in an `if`. Sounds something that can break very easily over time.

Possibly it is better to put those lines in a function, and call that where you expect? (in 2 situations if I read it right? `MC_HOVER` and an `else` of the `if`?)

Reply:
I kind of agree. Initially I just had an if statement all the way at the end checking excluding the specific situations where we shouldn't reset the values. However, the switch statement before this one [line 2870](https://github.com/alexanderweiss/OpenTTD/blob/8b6d2d470ef09e298acbc9358bab6d360ed494bf/src/window.cpp#L2870) also uses early returns, so it seemed sensible to copy the pattern.

As I suggested in a comment on my pull request it may be valuable to actually clean up the whole function as there's more stuff that isn't entirely logical, such as the check for w == NULL on line 2903, which can never be true as there's an early return on line 2852 if w == NULL. I'm not sure if cleaning it all up is appreciated though?

Reply:
Cleaning up can always be a good thing, but possibly not for this PR (maybe another?). But yeah, I understand why you did this. Okay, leave it as it is :) Now we just have to wait for @andythenorth  :D